sycl/test-e2e/Annotated_usm/annotated_usm.cpp

```diff
@@ -0,0 +1,173 @@
+// RUN: %{build} -o %t.out
+// RUN: %{run} %t.out
+
+// E2E tests for annotated USM allocation functions
+
+#include <sycl/sycl.hpp>
+
+#include <complex>
+#include <numeric>
+
+// clang-format on
+
+using namespace sycl::ext::oneapi::experimental;
+using namespace sycl::ext::intel::experimental;
+using alloc = sycl::usm::alloc;
+
+// Single test instance consisting of (i) calling malloc function `f` (ii)
+// checking usm location of returned pointer  (iii) calling free
+#define TEST_USM_KIND(f, kind, args...)                                        \
+  {                                                                            \
+    auto ap = f(args);                                                         \
+    assert(ap.get() != nullptr &&                                              \
+           sycl::get_pointer_type(ap.get(), Ctx) == kind);                     \
+    free(ap, q);                                                               \
+  }
+
+// Single test instance consisting of (i) calling malloc function `f` (ii)
+// checking alignment Note that the annotated_ptr is not freed in alignment
+// tests to avoid the same base address being returned every time
+#define TEST_ALIGN(f, align, args...)                                          \
+  {                                                                            \
+    auto ap = f(args);                                                         \
+    assert(ap.get() != nullptr && ((uintptr_t)ap.get() % align) == 0);         \
+  }
+
+// Single test instance consisting of (i) calling malloc function `f` (ii)
+// checking return is nullptr (iii) calling free
+#define TEST_ALIGN_NULL_EXPECTED(f, args...)                                   \
+  {                                                                            \
+    auto ap = f(args);                                                         \
+    assert(ap.get() == nullptr);                                               \
+    free(ap, q);                                                               \
+  }
+
+constexpr int N = 10;
+
+template <typename T> void testUsmKind(sycl::queue &q) {
+  const sycl::context &Ctx = q.get_context();
+  auto dev = q.get_device();
+
+  TEST_USM_KIND(malloc_annotated<T>, alloc::device, N, q, alloc::device);
+  TEST_USM_KIND(malloc_annotated, alloc::device, N, q, alloc::device);
```

Comment:
Can you add these tests as well, so we have complete coverage:

```
TEST_USM_KIND(aligned_alloc_annotated<T>, alloc::device, 1, N, q, alloc::device);
TEST_USM_KIND(aligned_alloc_annotated, alloc::device, 1, N, q, alloc::device);
```

And then these as well:

```
TEST_USM_KIND(malloc_annotated<T>, alloc::host, N, q, alloc::host);
TEST_USM_KIND(malloc_annotated, alloc::host, N, q, alloc::host);
```

Reply:
test cases added